toolchain/check/facet_type.cpp

```diff
@@ -253,54 +256,68 @@ auto AllocateFacetTypeImplWitness(Context& context,
+        lhs_access->witness_id);
+    auto rhs_lookup = context.insts().TryGetAs<SemIR::LookupImplWitness>(
+        rhs_access->witness_id);
+    if (lhs_lookup && rhs_lookup) {
+      if (lhs_access->index != rhs_access->index) {
+        return lhs_access->index <=> rhs_access->index;
+      }
+      return lhs_lookup->query_specific_interface_id.index <=>
+             rhs_lookup->query_specific_interface_id.index;
+    }
   }
-  if (auto bind_symbolic_name =
-          context.insts().TryGetAs<SemIR::BindSymbolicName>(inst_id)) {
-    const auto& bind_name =
-        context.entity_names().Get(bind_symbolic_name->entity_name_id);
-    return bind_name.name_id == SemIR::NameId::PeriodSelf;
+
+  if (context.constant_values().GetConstantInstId(lhs_id) ==
+      context.constant_values().GetConstantInstId(rhs_id)) {
+    return std::weak_ordering::equivalent;
   }
-  return false;
+  // Try return things in the order they appear in the code by using the
+  // non-canonicalized id.
+  return lhs_id.index <=> rhs_id.index;
 }
 
-// If `inst_id` is an `ImplWitnessAccess` into `.Self`, return it. Otherwise,
-// returns `nullopt`.
-static auto TryGetImplWitnessAccessOfPeriodSelf(Context& context,
-                                                SemIR::InstId inst_id)
-    -> std::optional<SemIR::ImplWitnessAccess> {
-  auto lhs_access = context.insts().TryGetAs<SemIR::ImplWitnessAccess>(inst_id);
-  if (!lhs_access) {
-    return std::nullopt;
-  }
-  auto lhs_lookup = context.insts().TryGetAs<SemIR::LookupImplWitness>(
-      lhs_access->witness_id);
-  if (!lhs_lookup) {
-    return std::nullopt;
-  }
-  if (!IsPeriodSelf(context, lhs_lookup->query_self_inst_id)) {
-    return std::nullopt;
-  }
-  return lhs_access;
+// Sort and dedupe the rewrite constraints, with accesses to the same associated
+// constants being treated as equivalent.
+static auto SortAndDedupeRewriteConstraints(
+    Context& context,
+    llvm::SmallVector<SemIR::FacetTypeInfo::RewriteConstraint>& rewrites) {
+  auto cmp = [&](const SemIR::FacetTypeInfo::RewriteConstraint& a,
+                 const SemIR::FacetTypeInfo::RewriteConstraint& b) {
+    auto lhs = CompareFacetTypeConstraintValues(context, a.lhs_id, b.lhs_id);
+    if (lhs != std::weak_ordering::equivalent) {
+      return lhs == std::weak_ordering::less;
+    }
+    auto rhs = CompareFacetTypeConstraintValues(context, a.rhs_id, b.rhs_id);
+    return rhs == std::weak_ordering::less;
+  };
+  llvm::stable_sort(rewrites, cmp);
+
+  auto eq = [&](const SemIR::FacetTypeInfo::RewriteConstraint& a,
+                const SemIR::FacetTypeInfo::RewriteConstraint& b) {
+    auto lhs = CompareFacetTypeConstraintValues(context, a.lhs_id, b.lhs_id);
+    if (lhs != std::weak_ordering::equivalent) {
+      return false;
+    }
+    auto rhs = CompareFacetTypeConstraintValues(context, a.rhs_id, b.rhs_id);
+    return rhs == std::weak_ordering::equivalent;
+  };
+  rewrites.erase(llvm::unique(rewrites, eq), rewrites.end());
```

Comment:
Just a thought (it's unfortunate these don't take a weak ordering directly):

```suggestion
  auto cmp = [&](const SemIR::FacetTypeInfo::RewriteConstraint& a,
                 const SemIR::FacetTypeInfo::RewriteConstraint& b) {
    auto lhs = CompareFacetTypeConstraintValues(context, a.lhs_id, b.lhs_id);
    if (lhs != std::weak_ordering::equivalent) {
      return lhs;
    }
    return CompareFacetTypeConstraintValues(context, a.rhs_id, b.rhs_id);
  };

  auto sort = [&](const SemIR::FacetTypeInfo::RewriteConstraint& a,
                  const SemIR::FacetTypeInfo::RewriteConstraint& b) {
    return cmp(a, b) == std::weak_ordering::less;
  }
  llvm::stable_sort(rewrites, sort);

  auto eq = [&](const SemIR::FacetTypeInfo::RewriteConstraint& a,
                const SemIR::FacetTypeInfo::RewriteConstraint& b) {
    return cmp(a, b) == std::weak_ordering::equivalent;
  }
  rewrites.erase(llvm::unique(rewrites, eq), rewrites.end());
```

Reply:
> it's unfortunate these don't take a weak ordering directly

Yeah. :(

Done (slightly different names)